src/app/directives/datatable/pagination.directive.ts

```diff
@@ -0,0 +1,267 @@
+})
+export class DatatablePaginationDirective<Model extends AbstractModel>
+  extends withUnsubscribe()
+  implements OnInit
+{
+  /**
+   * Base api filters for table. On trigger, this will update the table to
+   * match the new filters
+   */
+  @Input("paginationFilters$") public filters$: Observable<Filters<Model>>;
+  /** Base api filters for table */
+  @Input("paginationFilters") public filters: Filters<Model>;
+  /**
+   * Get models for table. This will be fed the filters for the current page of
+   * the table, and should output the matching models
+   */
+  @Input("paginationGetModels") public getModels: (
+    filters: Filters<Model>
+  ) => Observable<Model[]>;
+  /**
+   * Shared with ngx-datatable columns input, this should include the key (used
+   * for filter requests, and should be a filterable model key) and name of the
+   * column (used by ngx-datable for column names)
+   */
+  @Input() public columns: Column<Model>[];
+
+  /**
+   * Observable which outputs each table row. Each row value will be the model
+   * retrieved from the api request
+   */
+  private rows$: Observable<any>;
+  /**
+   * Observable tracking the total number of models for the current base filter
+   */
+  private total$: Observable<number>;
+  /** Observable tracking when the table is loading new data */
+  private loading$ = new BehaviorSubject<boolean>(false);
+  /**
+   * Observable tracking any changes to the current page, or sort, of the table
+   */
+  private pageAndSort$ = new BehaviorSubject<PageAndSort<Model>>({ page: 0 });
+
+  public constructor(@Host() private datatable: DatatableComponent) {
+    super();
+  }
+
+  public ngOnInit(): void {
+    // Convert basic filters to observable
+    if (this.filters && !this.filters$) {
+      this.filters$ = new BehaviorSubject(this.filters);
+    }
+
+    // Get the latest list of models whenever a change occurs to the page,
+    // sorting, or base filters
+    const models$ = combineLatest([this.pageAndSort$, this.filters$]).pipe(
+      // Show loading animation during request
+      tap((): void => this.loading$.next(true)),
+      // Combine base filter, paging, and sorting
+      map(
+        ([pageAndSort, filters]): Filters<Model> => ({
+          ...filters,
+          sorting: pageAndSort.sort ?? undefined,
+          // ngx-datatable uses page 0 as the first page, but the api uses page 1
+          paging: { page: pageAndSort.page + 1 },
+        })
+      ),
+      switchMap((filters): Observable<Model[]> => this.getModels(filters)),
+      // Clear loading animation
+      tap((): void => this.loading$.next(false))
+    );
+
+    // Convert models into rows by creating an object where each key is the
+    // column prop, and the value is the model
+    const convertToRow = (model: Model): any =>
+      this.columns
+        .map((column) => ({ [column.prop]: model }))
+        .reduce((previous, current) => ({ ...previous, ...current }), {});
```

Comment:
two notes here:

- repeated spread and merge into a object is a bad idea.
- you've got a O(2N) operation here when it could be O(N). As in looping twice through the set

You could use reduce just once  with the body changed to 
```
(previous, current) => { previous[current.prop] = model; previous}
```

Or use something like this to create the dictionary values all at one:
```
Object.fromEntries( [1,2,3].map((value) => [String.fromCharCode(value + 65), value]) )
```